### PR TITLE
feat: support for route permissions based on right codes GENC-469

### PR DESCRIPTION
**Pull request by SzymonZur**
[Review Guidance](https://www.notion.so/genesisglobal/Platform-Code-Review-Process-ace93ba760cc4563b0dfb712e2a88d8a)

📓 &nbsp; **Related JIRA**

[GENC-469](https://genesisglobal.atlassian.net/browse/GENC-469)

🤔 &nbsp; **What does this PR do?**

- As titled 

🚀 &nbsp; **Where should the reviewer start?**

Add file / directory pointers.

📑 &nbsp; **How should this be tested?**

### Defaults

```
rm -rf blankappseedtest && npx -y @genesislcap/genx@latest init blankappseedtest -x --ref sz/GENC-469 --no-npm
```

### Route and CSV parameter handling test

```
rm -rf blankappseedtest && npx -y @genesislcap/genx@latest init blankappseedtest --ref sz/GENC-469 --no-npm -x --routes '[{"name":"Home","tiles":[{"title":"Entity manager","type":"entity-manager","config":{ "modalPosition": "centre", "sizeColumnsToFit": true, "enableSearchBar": true, "resourceName":"ALL_POSITIONS","title":"My Positions","updateEvent":"EVENT_COUNTERPARTY_MODIFY","deleteEvent":"EVENT_COUNTERPARTY_DELETE","createEvent":"EVENT_COUNTERPARTY_INSERT", "createFormUiSchema": {"type":"VerticalLayout","elements":[{"type":"Control","label":"Inline create schema - main contact","scope":"#/properties/MAIN_CONTACT"},{"type":"Control","label":"Issuer Name - Local Schema","scope":"#/properties/ISSUER_NAME","options":{"readonly":true}},{"type":"Control","label":"Price","scope":"#/properties/PRICE"},{"type":"Control","scope":"#/properties/COUNTERPARTY","options":{"allOptionsResourceName":"ALL_COUNTERPARTYS","valueField":"COUNTERPARTY_ID","labelField":"COUNTERPARTY_ID"}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"isPassword":true}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"textarea":true}}]}, "updateFormUiSchema": {"type":"VerticalLayout","elements":[{"type":"Control","label":"Inline update schema - main contact","scope":"#/properties/MAIN_CONTACT"},{"type":"Control","label":"Issuer Name - Local Schema","scope":"#/properties/ISSUER_NAME","options":{"readonly":true}},{"type":"Control","label":"Price","scope":"#/properties/PRICE"},{"type":"Control","scope":"#/properties/COUNTERPARTY","options":{"allOptionsResourceName":"ALL_COUNTERPARTYS","valueField":"COUNTERPARTY_ID","labelField":"COUNTERPARTY_ID"}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"isPassword":true}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"textarea":true}}]}, "columns": [{"field":"INSTRUMENT_NAME","headerName":"Instrument Name"},{"field":"VALUE","headerName":"Inline coldef - VALUE"},{"field":"QUANTITY","headerName":"Quantity"},{"field":"PNL","headerName":"PNL"}] }},{"title":"Grid","type":"grid-pro","config":{"resourceName":"ALL_TRADES", "deferredGridOptions": {"columnDefs":[{"field":"INSTRUMENT_NAME","headerName":"Instrument Name"},{"field":"VALUE","headerName":"Inline coldef - VALUE"},{"field":"QUANTITY","headerName":"Quantity"},{"field":"PNL","headerName":"PNL"}]} }},{"title":"Form","type":"smart-form","config":{"resourceName":"EVENT_COUNTERPARTY_INSERT", "uischema": {"type":"VerticalLayout","elements":[{"type":"Control","label":"Inline form schema - main contact","scope":"#/properties/MAIN_CONTACT"},{"type":"Control","label":"Issuer Name - Local Schema","scope":"#/properties/ISSUER_NAME","options":{"readonly":true}},{"type":"Control","label":"Price","scope":"#/properties/PRICE"},{"type":"Control","scope":"#/properties/COUNTERPARTY","options":{"allOptionsResourceName":"ALL_COUNTERPARTYS","valueField":"COUNTERPARTY_ID","labelField":"COUNTERPARTY_ID"}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"isPassword":true}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"textarea":true}}]} }}]}, {"name":"Realtime Dashboard","tiles":[{"title":"Entity manager tile","type":"entity-manager","config":{"resourceName":"ALL_COUNTERPARTYS","title":"Counterparty Management","updateEvent":"EVENT_COUNTERPARTY_MODIFY","deleteEvent":"EVENT_COUNTERPARTY_DELETE","createEvent":"EVENT_COUNTERPARTY_INSERT"}},{"title":"Form tile","type":"smart-form","config":{"resourceName":"EVENT_COUNTERPARTY_INSERT"}}]},{"name":"Analytics","tiles":[{"title":"Grid Tile","type":"grid-pro","config":{"resourceName":"ALL_POSITIONS"}},{"title":"Charts Tile 1","type":"chart","config":{"type":"line","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_NAME","yField":"VALUE"}},{"title":"Charts Tile 2","type":"chart","config":{"type":"pie","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_NAME","yField":"VALUE"}},{"title":"Charts Tile 3","type":"chart","config":{"type":"column","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_ID","yField":"VALUE"}}]}]' --csv '[{"name": "trade", "fields": ["a", "B"]}, {"name": "position", "fields": ["id", "TYPE"]} ]' --apiHost 'wss://public-foundation.genesislab.global/gwf/' --no-shell && cd blankappseedtest/client && npm run bootstrap && npm run dev
```

### Permissions example

```
rm -rf blankappseedtest && npx -y @genesislcap/genx@latest init blankappseedtest --ref sz/GENC-469 --no-npm -x --routes '[
  {
    "name": "Home",
    "permissions": {
      "viewRight": null   
    },
    "tiles": [
      {
        "title": "Entity manager",
        "type": "entity-manager",
        "config": {
          "permissions": {
            "viewRight": "viewEM",    
            "updateRight": "updateEM"    
          },
          "modalPosition": "centre",
          "sizeColumnsToFit": true,
          "enableSearchBar": true,
          "resourceName": "ALL_POSITIONS",
          "title": "My Positions",
          "updateEvent": "EVENT_COUNTERPARTY_MODIFY",
          "deleteEvent": "EVENT_COUNTERPARTY_DELETE",
          "createEvent": "EVENT_COUNTERPARTY_INSERT",
          "createFormUiSchema": {
            "type": "VerticalLayout",
            "elements": [
              {
                "type": "Control",
                "label": "Inline create schema - main contact",
                "scope": "#/properties/MAIN_CONTACT"
              },
              {
                "type": "Control",
                "label": "Issuer Name - Local Schema",
                "scope": "#/properties/ISSUER_NAME",
                "options": {
                  "readonly": true
                }
              },
              {
                "type": "Control",
                "label": "Price",
                "scope": "#/properties/PRICE"
              },
              {
                "type": "Control",
                "scope": "#/properties/COUNTERPARTY",
                "options": {
                  "allOptionsResourceName": "ALL_COUNTERPARTYS",
                  "valueField": "COUNTERPARTY_ID",
                  "labelField": "COUNTERPARTY_ID"
                }
              },
              {
                "type": "Control",
                "label": "Password",
                "scope": "#/properties/PASSWORD",
                "options": {
                  "isPassword": true
                }
              },
              {
                "type": "Control",
                "label": "Password",
                "scope": "#/properties/PASSWORD",
                "options": {
                  "textarea": true
                }
              }
            ]
          },
          "updateFormUiSchema": {
            "type": "VerticalLayout",
            "elements": [
              {
                "type": "Control",
                "label": "Inline update schema - main contact",
                "scope": "#/properties/MAIN_CONTACT"
              },
              {
                "type": "Control",
                "label": "Issuer Name - Local Schema",
                "scope": "#/properties/ISSUER_NAME",
                "options": {
                  "readonly": true
                }
              },
              {
                "type": "Control",
                "label": "Price",
                "scope": "#/properties/PRICE"
              },
              {
                "type": "Control",
                "scope": "#/properties/COUNTERPARTY",
                "options": {
                  "allOptionsResourceName": "ALL_COUNTERPARTYS",
                  "valueField": "COUNTERPARTY_ID",
                  "labelField": "COUNTERPARTY_ID"
                }
              },
              {
                "type": "Control",
                "label": "Password",
                "scope": "#/properties/PASSWORD",
                "options": {
                  "isPassword": true
                }
              },
              {
                "type": "Control",
                "label": "Password",
                "scope": "#/properties/PASSWORD",
                "options": {
                  "textarea": true
                }
              }
            ]
          },
          "columns": [
            {
              "field": "INSTRUMENT_NAME",
              "headerName": "Instrument Name"
            },
            {
              "field": "VALUE",
              "headerName": "Inline coldef - VALUE"
            },
            {
              "field": "QUANTITY",
              "headerName": "Quantity"
            },
            {
              "field": "PNL",
              "headerName": "PNL"
            }
          ]
        }
      },
      {
        "title": "Grid",
        "type": "grid-pro",
        "config": {
          "resourceName": "ALL_TRADES",
          "permissions": {
            "viewRight": "viewTrades" 
          },
          "deferredGridOptions": {
            "columnDefs": [
              {
                "field": "INSTRUMENT_NAME",
                "headerName": "Instrument Name"
              },
              {
                "field": "VALUE",
                "headerName": "Inline coldef - VALUE"
              },
              {
                "field": "QUANTITY",
                "headerName": "Quantity"
              },
              {
                "field": "PNL",
                "headerName": "PNL"
              }
            ]
          }
        }
      },
      {
        "title": "Form",
        "type": "smart-form",
        "config": {
          "resourceName": "EVENT_COUNTERPARTY_INSERT",
          "permissions": {
            "updateRight": "updateCounterparties" 
          },
          "uischema": {
            "type": "VerticalLayout",
            "elements": [
              {
                "type": "Control",
                "label": "Inline form schema - main contact",
                "scope": "#/properties/MAIN_CONTACT"
              },
              {
                "type": "Control",
                "label": "Issuer Name - Local Schema",
                "scope": "#/properties/ISSUER_NAME",
                "options": {
                  "readonly": true
                }
              },
              {
                "type": "Control",
                "label": "Price",
                "scope": "#/properties/PRICE"
              },
              {
                "type": "Control",
                "scope": "#/properties/COUNTERPARTY",
                "options": {
                  "allOptionsResourceName": "ALL_COUNTERPARTYS",
                  "valueField": "COUNTERPARTY_ID",
                  "labelField": "COUNTERPARTY_ID"
                }
              },
              {
                "type": "Control",
                "label": "Password",
                "scope": "#/properties/PASSWORD",
                "options": {
                  "isPassword": true
                }
              },
              {
                "type": "Control",
                "label": "Password",
                "scope": "#/properties/PASSWORD",
                "options": {
                  "textarea": true
                }
              }
            ]
          }
        }
      }
    ]
  },
  {
    "name": "Realtime Dashboard",
    "permissions": {
      "viewRight": "viewDashboard" 
    },
    "tiles": [
      {
        "title": "Entity manager tile",
        "type": "entity-manager",
        "config": {
          "resourceName": "ALL_COUNTERPARTYS",
          "permissions": {
            "viewRight": null,
            "updateRight": "updateDashboard"
          },
          "title": "Counterparty Management",
          "updateEvent": "EVENT_COUNTERPARTY_MODIFY",
          "deleteEvent": "EVENT_COUNTERPARTY_DELETE",
          "createEvent": "EVENT_COUNTERPARTY_INSERT"
        }
      },
      {
        "title": "Form tile",
        "type": "smart-form",
        "config": {
          "permissions": {
            "updateRight": null
          },
          "resourceName": "EVENT_COUNTERPARTY_INSERT"
        }
      }
    ]
  },
  {
    "name": "Analytics",
    "permissions": {
      "viewRight": "viewAnalytics" 
    },
    "tiles": [
      {
        "title": "Grid Tile",
        "type": "grid-pro",
        "config": {
          "permissions": {
            "viewRight": null
          },
          "resourceName": "ALL_POSITIONS"
        }
      },
      {
        "title": "Charts Tile 1",
        "type": "chart",
        "config": {
          "permissions": {
            "viewRight": null
          },
          "type": "line",
          "resourceName": "ALL_POSITIONS",
          "xField": "INSTRUMENT_NAME",
          "yField": "VALUE"
        }
      },
      {
        "title": "Charts Tile 2",
        "type": "chart",
        "config": {
          "permissions": {
            "viewRight": null
          },
          "type": "pie",
          "resourceName": "ALL_POSITIONS",
          "xField": "INSTRUMENT_NAME",
          "yField": "VALUE"
        }
      },
      {
        "title": "Charts Tile 3",
        "type": "chart",
        "config": {
          "permissions": {
            "viewRight": null
          },
          "type": "column",
          "resourceName": "ALL_POSITIONS",
          "xField": "INSTRUMENT_ID",
          "yField": "VALUE"
        }
      }
    ]
  }
]' --csv '[{"name": "trade", "fields": ["a", "B"]}, {"name": "position", "fields": ["id", "TYPE"]} ]' --apiHost 'wss://public-foundation.genesislab.global/gwf/' --no-shell && cd blankappseedtest/client && npm run bootstrap && npm run dev
```

✅ &nbsp; **Checklist**

- [ ] I have tested my changes.
- [ ] I have added tests for my changes.
- [ ] I have updated the project documentation to reflect my changes.

### CODE REVIEW

**Comment by cistov** [comment]
TS drama in the build:

```
ERROR in /tmp/testapp/client/src/routes/config.ts
./src/routes/config.ts 85:12-26
[tsl] ERROR in /tmp/testapp/client/src/routes/config.ts(85,13)
      TS2322: Type '{ title: string; icon: { name: "cog"; variant: "solid"; }; permission: string; }' is not assignable to type 'FoundationRouteNavItem'.
  Object literal may only specify known properties, and 'permission' does not exist in type 'FoundationRouteNavItem'.
 @ ./src/routes/index.ts 1:0-25 1:0-25
 @ ./src/main/main.ts 9:0-45 94:11-27
 @ ./src/main/index.ts 2:0-23 2:0-23
 @ ./src/index.ts 1:0-29 1:0-29
 @ ./src/index.federated.ts
 @ container entry ./index[0]
```

File: client/src/routes/config.ts [hunk 1]
```diff
@@ -136,4 +141,25 @@ export class MainRouterConfig extends FoundationRouterConfiguration<LoginSetting
...
+        this.redirectIfNotPermitted(settings, phase);
+      }
+      return authenticated;
+    });
+  }
+
+  private redirectIfNotPermitted(settings: LoginSettings, phase: NavigationPhase) {
+    const { path } = phase.route.endpoint;
+    if (settings?.isPermitted && !settings.isPermitted()) {
+      const route = 'not-permitted';
```
Comment (cistov) [inline client/src/routes/config.ts:158]: If we make a typo in the route here, it'll just silently fail. We can assign this route/path to a variable and reuse e.g. not-permitted route component could provide an export with default route.

Comment (SzymonZur) [reply client/src/routes/config.ts]: [f2b3bd6](https://github.com/genesiscommunitysuccess/blank-app-seed/pull/255/commits/f2b3bd671bdca43dc2469035ec93c92559367164)

**Comment by cistov** [comment]
Looking good:

Screenshot 2024-06-14 at 12 51 42

**Review by cistov — APPROVED** [review]

**Pull request merged**